### ISSUE TITLE
Convert deprecation warnings to errors on CI

### ISSUE THREAD

**Issue by jakirkham** [rapidsai/build-planning#26]
To make it easier to catch and fix deprecations in RAPIDS projects, it is worth considering converting deprecation warnings to errors on CI. That way deprecations fail loudly and we are able to catch and address them quickly. Alternatively we can use that opportunity to tighten our dependencies and flag the deprecation for follow up when we are ready.

**Comment by mroeschke** [comment]
Progress on this front:

- [x] cudf
- [ ] dask_cudf https://github.com/rapidsai/cudf/pull/15634
- [x] cuml https://github.com/rapidsai/cuml/pull/5877
- [x] cuspatial https://github.com/rapidsai/cuspatial/pull/1360
- [ ] cuproj https://github.com/rapidsai/cuspatial/pull/1379
- [x] rmm https://github.com/rapidsai/rmm/pull/1551
- [x] kviko https://github.com/rapidsai/kvikio/pull/375
- [x] cuvs https://github.com/rapidsai/cuvs/pull/102
- [ ] cucim https://github.com/rapidsai/cucim/pull/734
- [ ] cuxfilter https://github.com/rapidsai/cuxfilter/pull/595
- [ ] custreamz
- [ ] cugraph

Let me know if I missed any projects

**Comment by jameslamb** [comment]
This is awesome! I'm also happy to see the consolidation into `pyproject.toml` and eliminating `pytest.ini`.

I see one that might have been missed... what about `raft`?

**Comment by mroeschke** [comment]
Did I miss any RAPIDS repos not covered in https://github.com/rapidsai/build-planning/issues/26#issuecomment-2091560816?

Additionally, to ensure that future RAPIDS repos have this change, is there like a template repo where we could enable this by default?

**Comment by jakirkham** [comment]
Thanks Matt for handling this for Python! 🙏

Is there anything we still need to do on the C++ side?

**Comment by mroeschke** [comment]
> Is there anything we still need to do on the C++ side?

Yeah I think all RAPIDS libraries should have `-Wdeprecated-declarations` enabled by default e.g. https://github.com/rapidsai/cudf/issues/14819.

**Comment by vyasr** [comment]
Not all RAPIDS libraries have this on by default yet. In particular, I think we'll have to update cuml/cugraph. This may still require some real C++ work to update the API calls though. I don't know if you want to track that effort in this issue, close this out and open repo-specific ones for the remaining cases, or some combination of the two. Most likely you'll want help from C++ library devs to address the outstanding issues there, though.

**Comment by jakirkham** [comment]
Think it is find to use this as an umbrella issue for the work needed to reach this goal

**Comment by jakirkham** [comment]
Another piece we may want to capture here is converting warnings during builds to errors

This may relevant to stay on top of things like...

* `scikit-build-core` changes: https://github.com/rapidsai/build-planning/issues/58
* `conda` package clobbering issues: https://github.com/rapidsai/build-planning/issues/54

...and possibly other use cases

**Comment by harrism** [comment]
RMM still has `-Wno-error=deprecated-declarations` in its test and benchmark cmakelists.txt. I'll look at fixing that.
